File: sdk-source/objc/Aegis/AegisClient.m
#import "AegisClient.h"
#import <CommonCrypto/CommonDigest.h>

@interface AegisClient ()
@property (nonatomic, strong) AegisOptions *options;
@property (nonatomic, copy) NSString *baseUrl;
@property (nonatomic, strong) NSURLSession *session;
@property (nonatomic, strong) dispatch_source_t heartbeatTimer;
@property (nonatomic, strong) dispatch_queue_t stateQueue;
@property (nonatomic, copy, nullable) NSString *token;
@end

@implementation AegisClient

+ (NSString *)sdkVersion {
    return @"1.0.0";
}

- (instancetype)initWithOptions:(AegisOptions *)options {
    NSParameterAssert(options.baseUrl.length > 0);
    NSParameterAssert(options.appKey.length > 0);

    if ((self = [super init])) {
        _options = options;
        _baseUrl = [options.baseUrl hasSuffix:@"/"] ? [options.baseUrl substringToIndex:options.baseUrl.length - 1]
                                                    : options.baseUrl;
        _hardwareId = options.hwid.length > 0 ? [options.hwid copy] : [AegisClient hardwareIdentifier];
        _stateQueue = dispatch_queue_create("io.aegis.sdk.state", DISPATCH_QUEUE_SERIAL);

        NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
        configuration.timeoutIntervalForRequest = options.timeout;
        _session = [NSURLSession sessionWithConfiguration:configuration];
    }
    return self;
}

+ (NSString *)hardwareIdentifier {
    NSProcessInfo *info = [NSProcessInfo processInfo];
    NSString *facts = [@[
        info.hostName ?: @"unknown",
        info.operatingSystemVersionString ?: @"",
        @(info.processorCount).stringValue,
        info.environment[@"USER"] ?: @"",
    ] componentsJoinedByString:@"|"];

    NSData *data = [facts dataUsingEncoding:NSUTF8StringEncoding];
    unsigned char digest[CC_SHA256_DIGEST_LENGTH];
    CC_SHA256(data.bytes, (CC_LONG)data.length, digest);

    NSMutableString *hex = [NSMutableString stringWithCapacity:CC_SHA256_DIGEST_LENGTH * 2];
    for (int i = 0; i < CC_SHA256_DIGEST_LENGTH; i++) {
        [hex appendFormat:@"%02x", digest[i]];
    }
    return hex;
}

- (nullable NSString *)sessionToken {
    __block NSString *current = nil;
    dispatch_sync(self.stateQueue, ^{ current = self.token; });
    return current;
}

- (void)useSession:(nullable NSString *)token {
    dispatch_async(self.stateQueue, ^{ self.token = [token copy]; });
}

#pragma mark - Transport

- (NSError *)errorWithCode:(NSString *)code message:(NSString *)message status:(NSInteger)status {
    return [NSError errorWithDomain:AegisErrorDomain
                               code:status
                           userInfo:@{ AegisErrorCodeKey: code, NSLocalizedDescriptionKey: message }];
}

- (void)request:(NSString *)endpoint body:(nullable NSDictionary *)body completion:(AegisCompletion)completion {
    [self performRequest:endpoint body:body attempt:0 completion:completion];
}

- (void)performRequest:(NSString *)endpoint
                  body:(nullable NSDictionary *)body
               attempt:(NSUInteger)attempt
            completion:(AegisCompletion)completion {
    NSString *path = [endpoint stringByTrimmingCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@"/"]];
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"%@/api/public/v1/%@", self.baseUrl, path]];
    if (url == nil) {
        completion(nil, [self errorWithCode:@"invalid_options" message:@"Invalid baseUrl." status:0]);
        return;
    }

    NSMutableDictionary *payload = [NSMutableDictionary dictionary];
    [body enumerateKeysAndObjectsUsingBlock:^(id key, id value, BOOL *stop) {
        if (value != nil && value != [NSNull null]) payload[key] = value;
    }];

    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"POST";
    request.timeoutInterval = self.options.timeout;
    [request setValue:@"application/json" forHTTPHeaderField:@"content-type"];
    [request setValue:[NSString stringWithFormat:@"aegis-objc-sdk/%@", AegisClient.sdkVersion]
   forHTTPHeaderField:@"user-agent"];
    [request setValue:self.options.appKey forHTTPHeaderField:@"x-app-key"];
    if (self.options.apiKey.length > 0) [request setValue:self.options.apiKey forHTTPHeaderField:@"x-api-key"];
    NSString *current = self.sessionToken;
    if (current.length > 0) [request setValue:current forHTTPHeaderField:@"x-session-token"];
    request.HTTPBody = [NSJSONSerialization dataWithJSONObject:payload options:0 error:nil];

    __weak typeof(self) weakSelf = self;
    NSURLSessionDataTask *task = [self.session dataTaskWithRequest:request
                                                 completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        __strong typeof(weakSelf) self = weakSelf;
        if (self == nil) return;

        NSInteger status = [response isKindOfClass:NSHTTPURLResponse.class] ? ((NSHTTPURLResponse *)response).statusCode : 0;
        BOOL retryable = (error != nil || status >= 500) && attempt < self.options.maxRetries;
        if (retryable) {
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.25 * (attempt + 1) * NSEC_PER_SEC)),
                           dispatch_get_global_queue(QOS_CLASS_UTILITY, 0), ^{
                [self performRequest:endpoint body:body attempt:attempt + 1 completion:completion];
            });
            return;
        }
        if (error != nil) {
            completion(nil, [self errorWithCode:@"network_error" message:error.localizedDescription status:0]);
            return;
        }

        NSDictionary *envelope = data.length > 0 ? [NSJSONSerialization JSONObjectWithData:data options:0 error:nil] : nil;
        if (![envelope isKindOfClass:NSDictionary.class]) {
            completion(nil, [self errorWithCode:@"invalid_response" message:@"Malformed API response." status:status]);
            return;
        }
        if ([envelope[@"success"] boolValue]) {
            NSDictionary *result = envelope[@"data"];
            completion([result isKindOfClass:NSDictionary.class] ? result : @{}, nil);
            return;
        }
        NSDictionary *failure = [envelope[@"error"] isKindOfClass:NSDictionary.class] ? envelope[@"error"] : @{};
        completion(nil, [self errorWithCode:failure[@"code"] ?: @"server_error"
                                    message:failure[@"message"] ?: @"Request failed."
                                     status:status]);
    }];
    [task resume];
}

- (void)storeSession:(NSDictionary *)data {
    NSDictionary *session = data[@"session"];
    if ([session isKindOfClass:NSDictionary.class] && [session[@"token"] isKindOfClass:NSString.class]) {
        [self useSession:session[@"token"]];
    }
}

#pragma mark - Application

- (void)initializeApplication:(AegisCompletion)completion {
    [self request:@"init" body:@{ @"version": self.options.version } completion:completion];
}

- (void)status:(AegisCompletion)completion { [self request:@"status" body:nil completion:completion]; }

- (void)appData:(AegisCompletion)completion { [self request:@"app/data" body:nil completion:completion]; }

- (void)downloads:(AegisCompletion)completion { [self request:@"downloads" body:nil completion:completion]; }

- (void)checkVersion:(nullable NSString *)version completion:(AegisCompletion)completion {
    [self request:@"version/check"
             body:@{ @"version": version ?: self.options.version, @"channel": self.options.channel }
       completion:completion];
}

#pragma mark - Authentication

- (void)registerUsername:(NSString *)username
                password:(NSString *)password
                   email:(nullable NSString *)email
              licenseKey:(nullable NSString *)licenseKey
              completion:(AegisCompletion)completion {
    __weak typeof(self) weakSelf = self;
    [self request:@"register"
             body:@{ @"username": username, @"password": password,
                     @"email": email ?: [NSNull null], @"license_key": licenseKey ?: [NSNull null],
                     @"hwid": self.hardwareId }
       completion:^(NSDictionary *data, NSError *error) {
        if (data != nil) [weakSelf storeSession:data];
        completion(data, error);
    }];
}

- (void)loginUsername:(NSString *)username password:(NSString *)password completion:(AegisCompletion)completion {
    __weak typeof(self) weakSelf = self;
    [self request:@"login"
             body:@{ @"username": username, @"password": password, @"hwid": self.hardwareId }
       completion:^(NSDictionary *data, NSError *error) {
        if (data != nil) [weakSelf storeSession:data];
        completion(data, error);
    }];
}

- (void)logout:(AegisCompletion)completion {
    __weak typeof(self) weakSelf = self;
    [self request:@"logout" body:nil completion:^(NSDictionary *data, NSError *error) {
        [weakSelf useSession:nil];
        completion(data, error);
    }];
}

- (void)heartbeat:(AegisCompletion)completion { [self request:@"heartbeat" body:nil completion:completion]; }

- (void)checkSession:(AegisCompletion)completion { [self request:@"session/check" body:nil completion:completion]; }

- (void)isAuthenticated:(void (^)(BOOL authenticated))completion {
    if (self.sessionToken.length == 0) {
        completion(NO);
        return;
    }
    [self checkSession:^(NSDictionary *data, NSError *error) {
        completion(error == nil && [data[@"valid"] boolValue]);
    }];
}

- (void)userData:(AegisCompletion)completion { [self request:@"user/data" body:nil completion:completion]; }

#pragma mark - Licensing

- (void)validateLicense:(NSString *)licenseKey completion:(AegisCompletion)completion {
    [self request:@"license/validate"
             body:@{ @"license_key": licenseKey, @"hwid": self.hardwareId }
       completion:completion];
}

- (void)activateLicense:(NSString *)licenseKey
               username:(nullable NSString *)username
             completion:(AegisCompletion)completion {
    [self request:@"license/activate"
             body:@{ @"license_key": licenseKey, @"hwid": self.hardwareId, @"username": username ?: [NSNull null] }
       completion:completion];
}

#pragma mark - Variables

- (void)variablesForScope:(nullable NSString *)scope
               licenseKey:(nullable NSString *)licenseKey
               completion:(AegisCompletion)completion {
    [self request:@"variables/get"
             body:@{ @"scope": scope ?: @"application", @"license_key": licenseKey ?: [NSNull null] }
       completion:completion];
}

- (void)setVariable:(NSString *)key
              value:(NSString *)value
              scope:(nullable NSString *)scope
         licenseKey:(nullable NSString *)licenseKey
         completion:(AegisCompletion)completion {
    [self request:@"variables/set"
             body:@{ @"scope": scope ?: @"user", @"key": key, @"value": value,
                     @"license_key": licenseKey ?: [NSNull null] }
       completion:completion];
}

- (void)triggerWebhook:(NSString *)event
               payload:(nullable NSDictionary *)payload
            completion:(AegisCompletion)completion {
    [self request:@"webhook/trigger" body:@{ @"event": event, @"payload": payload ?: @{} } completion:completion];
}

#pragma mark - Sessions

- (void)startHeartbeatWithInterval:(NSTimeInterval)interval onRevoked:(nullable void (^)(NSError *))onRevoked {
    [self stopHeartbeat];

    dispatch_queue_t queue = dispatch_get_global_queue(QOS_CLASS_UTILITY, 0);
    dispatch_source_t timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, queue);
    dispatch_source_set_timer(timer,
                              dispatch_time(DISPATCH_TIME_NOW, (int64_t)(interval * NSEC_PER_SEC)),
                              (uint64_t)(interval * NSEC_PER_SEC),
                              (uint64_t)(NSEC_PER_SEC));

    __weak typeof(self) weakSelf = self;
    dispatch_source_set_event_handler(timer, ^{
        [weakSelf heartbeat:^(NSDictionary *data, NSError *error) {
            if (error != nil && !error.aegis_isNetworkError) {
                [weakSelf stopHeartbeat];
                if (onRevoked != nil) onRevoked(error);
            }
        }];
    });
    self.heartbeatTimer = timer;
    dispatch_resume(timer);
}

- (void)stopHeartbeat {
    if (self.heartbeatTimer != nil) {
        dispatch_source_cancel(self.heartbeatTimer);
        self.heartbeatTimer = nil;
    }
}

- (void)dealloc {
    [self stopHeartbeat];
}

@end

File: sdk-source/swift/Sources/Aegis/AegisClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Client for the Aegis Authentication API.
/// Pure Foundation, no third-party dependencies. All calls are `async`.
public actor AegisClient {
    public static let sdkVersion = "1.0.0"

    private let options: AegisOptions
    private let baseUrl: String
    private let session: URLSession
    private var heartbeatTask: Task<Void, Never>?

    /// Machine identifier sent with authentication and licensing calls.
    public let hardwareId: String

    /// Current session token, or `nil` when signed out.
    public private(set) var sessionToken: String?

    public init(options: AegisOptions) {
        precondition(!options.baseUrl.isEmpty, "baseUrl is required.")
        precondition(!options.appKey.isEmpty, "appKey is required.")
        self.options = options
        self.baseUrl = options.baseUrl.hasSuffix("/") ? String(options.baseUrl.dropLast()) : options.baseUrl
        self.hardwareId = options.hwid ?? HardwareId.current()

        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = options.timeout
        self.session = URLSession(configuration: configuration)
    }

    /// Restores a session token persisted by the host application.
    public func useSession(_ token: String?) {
        sessionToken = token
    }

    /// Calls any endpoint and returns its `data` payload.
    @discardableResult
    public func request(_ endpoint: String, body: [String: Any] = [:]) async throws -> [String: Any] {
        let path = endpoint.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard let url = URL(string: "\(baseUrl)/api/public/v1/\(path)") else {
            throw AegisError(code: "invalid_options", message: "Invalid baseUrl.")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = options.timeout
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.setValue("aegis-swift-sdk/\(Self.sdkVersion)", forHTTPHeaderField: "user-agent")
        request.setValue(options.appKey, forHTTPHeaderField: "x-app-key")
        if let apiKey = options.apiKey { request.setValue(apiKey, forHTTPHeaderField: "x-api-key") }
        if let token = sessionToken { request.setValue(token, forHTTPHeaderField: "x-session-token") }
        request.httpBody = try JSONSerialization.data(withJSONObject: body.compactMapValues { $0 })

        var lastFailure: Error?
        for attempt in 0...options.maxRetries {
            do {
                let (data, response) = try await session.data(for: request)
                let status = (response as? HTTPURLResponse)?.statusCode ?? 0
                if status >= 500, attempt < options.maxRetries {
                    try? await Task.sleep(nanoseconds: UInt64(250_000_000 * (attempt + 1)))
                    continue
                }
                guard let envelope = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                    throw AegisError(code: "invalid_response", message: "Malformed API response.", status: status)
                }
                if envelope["success"] as? Bool == true {
                    return envelope["data"] as? [String: Any] ?? [:]
                }
                let error = envelope["error"] as? [String: Any] ?? [:]
                throw AegisError(
                    code: error["code"] as? String ?? "server_error",
                    message: error["message"] as? String ?? "Request failed.",
                    status: status
                )
            } catch let error as AegisError {
                throw error
            } catch {
                lastFailure = error
                if attempt < options.maxRetries {
                    try? await Task.sleep(nanoseconds: UInt64(250_000_000 * (attempt + 1)))
                }
            }
        }
        throw AegisError(code: "network_error", message: lastFailure?.localizedDescription ?? "Network request failed.")
    }

    // MARK: - Application

    /// Handshake. Call once before any other operation.
    @discardableResult
    public func initialize() async throws -> [String: Any] {
        try await request("init", body: ["version": options.version])
    }

    public func status() async throws -> [String: Any] { try await request("status") }

    public func appData() async throws -> [String: Any] { try await request("app/data") }

    /// Download information published for this application.
    public func downloads() async throws -> [String: Any] { try await request("downloads") }

    public func checkVersion(_ version: String? = nil) async throws -> [String: Any] {
        try await request("version/check", body: ["version": version ?? options.version, "channel": options.channel])
    }

    // MARK: - Authentication

    @discardableResult
    public func register(
        username: String,
        password: String,
        email: String? = nil,
        licenseKey: String? = nil
    ) async throws -> [String: Any] {
        let data = try await request("register", body: [
            "username": username, "password": password,
            "email": email as Any, "license_key": licenseKey as Any, "hwid": hardwareId,
        ])
        storeSession(data)
        return data
    }

    @discardableResult
    public func login(username: String, password: String) async throws -> [String: Any] {
        let data = try await request("login", body: ["username": username, "password": password, "hwid": hardwareId])
        storeSession(data)
        return data
    }

    public func logout() async throws {
        defer { sessionToken = nil }
        _ = try await request("logout")
    }

    @discardableResult
    public func heartbeat() async throws -> [String: Any] { try await request("heartbeat") }

    public func checkSession() async throws -> [String: Any] { try await request("session/check") }

    /// True when a token exists and the server still accepts it.
    public func isAuthenticated() async -> Bool {
        guard sessionToken != nil else { return false }
        return ((try? await checkSession())?["valid"] as? Bool) == true
    }

    public func userData() async throws -> [String: Any] { try await request("user/data") }

    // MARK: - Licensing

    public func validateLicense(_ licenseKey: String) async throws -> [String: Any] {
        try await request("license/validate", body: ["license_key": licenseKey, "hwid": hardwareId])
    }

    public func activateLicense(_ licenseKey: String, username: String? = nil) async throws -> [String: Any] {
        try await request("license/activate", body: [
            "license_key": licenseKey, "hwid": hardwareId, "username": username as Any,
        ])
    }

    // MARK: - Variables

    public func variables(scope: String = "application", licenseKey: String? = nil) async throws -> [String: Any] {
        try await request("variables/get", body: ["scope": scope, "license_key": licenseKey as Any])
    }

    @discardableResult
    public func setVariable(
        _ key: String,
        value: String,
        scope: String = "user",
        licenseKey: String? = nil
    ) async throws -> [String: Any] {
        try await request("variables/set", body: [
            "scope": scope, "key": key, "value": value, "license_key": licenseKey as Any,
        ])
    }

    @discardableResult
    public func triggerWebhook(event: String, payload: [String: Any] = [:]) async throws -> [String: Any] {
        try await request("webhook/trigger", body: ["event": event, "payload": payload])
    }

    // MARK: - Sessions

    /// Starts a background heartbeat; `onRevoked` fires once when the session dies.
    public func startHeartbeat(interval: TimeInterval = 60, onRevoked: @Sendable @escaping (String) -> Void = { _ in }) {
        stopHeartbeat()
        heartbeatTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
                guard let self, !Task.isCancelled else { return }
                do {
                    _ = try await self.heartbeat()
                } catch let error as AegisError {
                    await self.stopHeartbeat()
                    onRevoked(error.message)
                    return
                } catch {
                    // transient transport failure — retry on the next tick
                }
            }
        }
    }

    public func stopHeartbeat() {
        heartbeatTask?.cancel()
        heartbeatTask = nil
    }

    private func storeSession(_ data: [String: Any]) {
        if let session = data["session"] as? [String: Any], let token = session["token"] as? String {
            sessionToken = token
        }
    }
}